src/nprintml/label_aggregation.py

```diff
@@ -0,0 +1,233 @@
+a single sample and the labels for the samples are contained in a 'label' column
+in the dataframe
+
+"""
+
+import pathlib
+from abc import ABC, abstractmethod
+
+import numpy as np
+import pandas as pd
+
+
+class LabelAggregator(ABC):
+    '''
+    Label Aggregator provides general methods for working with nPrints,
+    including attaching labels, compressing nPrints, and generating new
+    columns for flattened nPrints.
+
+    Inheritors of this class are expected to implement generate_features() as
+    a custom method to provide different methods for aggregating features
+    '''
+    def __init__(self, npts, label_file):
+        self.npts = npts
+        self.label_file = label_file
+
+    @abstractmethod
+    def generate_features(self, compress=True, sample_size=1):
+        '''
+        Abstract method, expected to be implemented on a per-example label
+        aggregation method.
+        '''
+
+    def load_npt(self, npt):
+        '''
+        Load an nPrint, where the index column is *always* the 0th column.
+        '''
+        npt_df = pd.read_csv(npt, index_col=0)
+
+        return npt_df
+
+    def load_labels(self, labels):
+        '''
+        Load labels, which are expected to be in item,label column format where
+        item is the index.
+        '''
+        labels = pd.read_csv(labels, index_col=0)
+
+        return labels
+
+    def compress_npt(self, npt):
+        '''
+        Compress columns out of an nPrint that provide no predictive signal.
+        More specifically, this method drops *all* columns in a given nPrint
+        dataframe where each bit is the same value.
+        '''
+        nunique = npt.apply(pd.Series.nunique)
+        cols_to_drop = nunique[nunique == 1].index
+        npt = npt.drop(cols_to_drop, axis=1)
+
+        return npt
+
+    def attach_labels(self, npt_df, labels):
+        '''
+        Attach labels to a dataframe of nPrints, returning the labels that are
+        missing, the new dataframe, and the number of samples that were lost
+        '''
+        missing_labels = set(npt_df.index) - set(labels.index)
+        og_num_samples = npt_df.shape[0]
+        npt_df = npt_df.join(labels)
+        npt_df = npt_df.dropna(subset=['label'])
+        new_num_samples = npt_df.shape[0]
+
+        return npt_df, missing_labels, og_num_samples, new_num_samples
+
+    def get_flattened_columns(self, npt, sample_size):
+        '''
+        When we attach labels to more than one nPrint we need to create new
+        columns that essentially are the original columns multiplied by
+        the number of packets in each flattened sample.
+        '''
+        og_columns = list(npt.columns)
+        new_columns = []
+        for i in range(sample_size):
+            for column in og_columns:
+                new_columns.append('pkt_{0}_{1}'.format(i, column))
+
+        return new_columns
+
+
+class IndexLabelAggregator(LabelAggregator):
```

Comment:
The idea with the IndexLabelAggregator is that we can take any single nPrint that is supplied as normal and create features for it. We can leverage the Index (src IP, dst IP, etc.) and the user labels to have one unified set of functions as we don't care what the index is, just that it matches the supplied labels